### ISSUE TITLE
Eliminating IdWithAlt from the final output of Karatsuba

### ISSUE THREAD

**Issue by JasonGross** [mit-plv/fiat-crypto#195]
The first lines of Karatsuba before bounds analysis look like this:
```coq
λ '(H6, x9, x8, x7, x6, x5, x4, x3, (H13, x16, x15, x14, x13, x12, x11, x10))%core,
ℤ x17 = 0x0;
ℤ x18 = Op (Syntax.IdWithAlt Syntax.TZ Syntax.TZ Syntax.TZ)
          (x6 * x13 + H6 * H13 +
           ((x6 + H6) * (x13 + H13) - x6 * x13) *
           0x100000000000000000000000000000000000000000000000000000000L,
          x6 * x13 + H6 * H13 +
          (x6 * H13 + H6 * x13 + H6 * H13) *
          0x100000000000000000000000000000000000000000000000000000000L);
```
After bounds analysis, how do I eliminate `IdWithAlt`?  I see three options, order of easiest for me to hardest for me:
1. write a notation that makes `IdWithAlt` invisible 
2. change the pre-reified code, so that the two alternatives are `(x6 + H6) * (x13 + H13) - x6 * x13` and `x6 * H13 + H6 * x13 + H6 * H13` rather than these larger expressions; I can then have the simplifier push the bounds inwards one step, and make `IdWithAlt` go away at the same time, I hope.  (@jadephilipoom how much work would it be to fix this in the pre-reified code?)
3. write a pass that can push `IdWithAlt` inwards across identical subexpressions (this requires writing a version of the arithmetic simplifier that operates on `Named` syntax trees rather than on PHOAS, building the relevant machinery, and the writing the pusher---I expect this will take about half-a-week to a week)

@andres-erbsen Thoughts on this?

**Comment by andres-erbsen** [comment]
The notation is a fine kludge for testing. I am not sure I understand the other options -- shouldn't eliminating idwithalt involve proving that the sides are always the same?

**Comment by JasonGross** [comment]
> shouldn't eliminating idwithalt involve proving that the sides are always the same?

No.  Suppose we have
`L := ((x + y) * (z + w) - x * z)`
`R := (x * w + y * z + y * w)`
and we are looking at `id_with_alt L R`, and the bounds are `0 ≤ x, y, z, w ≤ B`.  Then we have the bounds:
- `0 ≤ (x + y) * (z + w) ≤ 4B²`
- `-B² ≤ L ≤ 4B²`
- `0 ≤ R ≤ 3B²`
- and hence `0 ≤ id_with_alt L R ≤ 3B²` by the side conditions of bounds analysis.

After word-size selection, `id_with_alt L R` returns the smallest word-size that fits `3B²` (e.g., `IdWithAlt TZ (TWord 7) (TWord 7)`, to denote that `L` is interpreted as a `Z` and `R` and the output fit in 128 bits).  However, the subtraction in `L` outputs a `Z`, because word-size selection doesn't know that it never underflows.  Since the interpretation of `id_with_alt L R` outputting a `TWord 7` is that you run `L` (which returns a `Z` and then cast to `TWord 7`, you can equivalently have the subtraction in `L` return `TWord 7` directly.  However, this by itself doesn't give a guarantee that none of the intermediate computations overflow or underflow, and we can only apply this reasoning to the outermost operation, so, e.g., if we put it in A-normal form, the inner operations might still be assigned to `Z` variables.  If we can make it so that the outermost operation is the only one that bounds analysis goes wrong on, then we can truly adjust the bounds, and get out code that doesn't involve `IdWithAlt` and has the tighter bounds.

**Comment by andres-erbsen** [comment]
By my understanding of what we agreed on in the in-person meeting where we decided to do id-with-alt, the issue here is that id-with-alt is in the wrong place (and the reflective pipeline is working as expected). We should either fix the synthesis to put id_with_alt right on the expressions that need manual bounds analysis, or revert it from the synthesis and insert it on synthesized code (I think I could do it interactively in not that much time). @jadephilipoom ?

**Comment by andres-erbsen** [comment]
I fixed both the mul-by-`2^224` and `id_with_alt` placement issue on https://github.com/mit-plv/fiat-crypto/tree/karatsuba-fix . However, it seems that we are totally missing a `reduce` instantiation for 448, so the integration test no longer builds after those fixes (previously, `goldilocks_mul` had the wrong type).

**Comment by jadephilipoom** [comment]
Inserted the `reduce`, but now reification fails with an anomaly. I have no idea what could be causing it-- @JasonGross, do you have any immediate ideas?

```
File "./src/Specific/IntegrationTestKaratsubaMul.v", line 79, characters 6-20:
Anomaly:
File "pretyping/constr_matching.ml", line 115, characters 15-21: Assertion failed.
Please report at http://coq.inria.fr/bugs/.
```

**Comment by JasonGross** [comment]
Yeah, this is the bug that I needed the really long name to fix.  The solution is most likely going to be either replacing an `_` with a dummy binder name, or eta-expanding a function, but I'm not sure how to easily track it down... you could set the reification debug level to something like 3 and look at the output to see how far it gets before it fails with the anomaly.  Or I can look into it once I get to a computer.

**Comment by JasonGross** [comment]
The anomaly is fixed by 30d8500
